### ISSUE TITLE
Overrides in s(CASP) version

### ISSUE THREAD

**Issue by Gauntlet173** [Lexpedite/blawx#130]
Because overrides are not native to s(CASP), we need to implement a library for them. This will likely only be feasible after we have a way of attributing the provenance of rules in legislation, so it may be dependent on the legal document features being implemented first. We could generate user-specified names for rules, as in v0, too, but that would almost certainly need to be replaced once the legal document features were implemented.

This will be based on the lpdat library I wrote for s(CASP), with some modifications, likely.

**Comment by Gauntlet173** [comment]
Adding the override block to the interface was done as part of adding legal documents to v1. So the task here is just to implement code generation for them and the back-end argumentation theory.

**Comment by Gauntlet173** [comment]
Completed those three in the defeasibility branch. Simplified the LDAP encoding to only use refutation, and change the override rules to all conclusions from each rule. The LDAP library is loaded in all queries. Code is generated for code identifiers and override blocks. Rule blocks are using the local rule, and queries are using the legally_holds.

Remaining for this issue:
* figure out how to give the user access to legally_holds, according_to, negated, and unwrapped statements in the question blocks, while keeping the interface as simple as possible.
* NLG updates for the explanations
* make the code generator create opposes statements for boolean attributes automatically.
* generate opposes statements implicitly for any logically negated statements?

Can be dealt with under a different issue:
* method of sub-dividing rule text in the code editor, or in the rule editor, at arbitrary levels of specificity, to allow for exceptions in sub-sections of text.
* update the LDAP library with defeated defeaters, defeat cycles, unresolved conflicts, etc.
* see if there is a way to simplify the overrides system in the argumentation theory without breaking things.
* scoping for defeasibility (the "in" block)

Order doesn't really matter on the remaining tasks for this issue.

**Comment by Gauntlet173** [comment]
Scoping is #181. The rest were created as issues.

**Comment by Gauntlet173** [comment]
the NLG updates are done in the defeasibility branch.

**Comment by Gauntlet173** [comment]
There is an additional issue with the existing library, which does only refutation, in that something does not hold legally unless there are at least two conclusions found. That will need to be resolved as a part of this issue.

**Comment by Gauntlet173** [comment]
The issue with refutations has been resolved in defeasibility by rewriting the library and changing how oppose code is generated. NLG for defeasibility exposed some weaknesses in the sCASP library, so the reasoner had to be re-written to get around those, too.

Currently, I need to write a function to extract variables from the query, and then we should have completely working NLG for defeasible conclusions.

Generating opposes for booleans and negations, and adding advanced defeasible statements and queries to the interface remain TBD on this issue. I'm considering having an "advanced" mode for some blocks accessible through a context menu that will make the defeasibility information explicit.

Also considering adding a "strict" defeasibility mode to the library and the interface.

**Comment by Gauntlet173** [comment]
Detecting variables, and oppose statements for boolean properties and negated category or attribute statements are now included. The tweety example has been updated to remove the extraneous overrules statement.

Now having a problem with the defeasibility library as a result of the changes to how oppose statements are generated. Need to resolve that and then deal with the interface.

**Comment by Gauntlet173** [comment]
The problems with the defeasibility library was being created by creating default opposes statements for logical negations. That isn't urgent, so I'm going to remove that for the time being and come back to it later as a different issue.

On the interface, I have decided to go with the simplest thing that could work, which is to make the use of the defeasibility interface explicit. So rules need an "according to" block, and queries need a "holds" block if that's what they are trying to check. The Tweety example has been updated in that way.

In order to simplify getting access to the combination of statements that you are likely to need, I'm going to add pre-assembled rules and queries. Getting the rule to include the current rule in the according to is going to require making the primary menu dynamic, so that is probably the last thing to do for this issue.
